envoy/server.go

```diff
@@ -0,0 +1,437 @@
+}
+
+// EnvoyClustersFromState genenerates a list of Envoy clusters from the
+// current Sidecar state
+func (s *Server) EnvoyClustersFromState() []cache.Resource {
+	s.state.RLock()
+	defer s.state.RUnlock()
+
+	// `s.state.ByService()` returns the list of service endpoints for each service.
+	// Since some services can expose multiple service ports, we need to create a
+	// separate cluster for each (service, servicePort) pair. If a service doesn't
+	// have any endpoints that are alive, we don't want to create a cluster for it.
+	//
+	// Note that in `EnvoyClustersFromState()` for the REST API we only need
+	// the first non-nil alive endpoint instance to construct the cluster
+	// because, in that case, SDS (now EDS) fetches the actual endpoints in a
+	// separate call.
+	var clusters []cache.Resource
+	clustersMap := make(map[string]*api.Cluster)
+	for svcName, svcEndpoints := range s.state.ByService() {
+		if len(svcEndpoints) < 1 {
+			continue
+		}
+
+		for _, svcEndpoint := range svcEndpoints {
+			if svcEndpoint == nil || !svcEndpoint.IsAlive() {
+				continue
+			}
+
+			for _, port := range svcEndpoint.Ports {
+				if port.ServicePort < 1 {
+					continue
+				}
+
+				envoyServiceName := SvcName(svcName, uint32(port.ServicePort))
+
+				if cluster, ok := clustersMap[envoyServiceName]; ok {
+					cluster.LoadAssignment.Endpoints[0].LbEndpoints =
+						append(cluster.LoadAssignment.Endpoints[0].LbEndpoints,
+							s.EnvoyServiceFromService(svcEndpoint, port.ServicePort)...)
+				} else {
+					envoyCluster := &api.Cluster{
+						Name: envoyServiceName,
+						// TODO: Are we happy with this timeout or should we make it configurable?
+						// It's the same as the one set in ConnectTimeoutMs in the REST API.
+						ConnectTimeout: &duration.Duration{Nanos: 500000000}, // 500ms
+						// TODO: This is the default policy and what we used for the REST API. Is this OK?
+						// LbPolicy: v2pb.Cluster_ROUND_ROBIN,
+						ClusterDiscoveryType: &api.Cluster_Type{Type: api.Cluster_STATIC}, // Use IPs only
+						ProtocolSelection:    api.Cluster_USE_CONFIGURED_PROTOCOL,
+						// Setting the endpoints here directly bypasses EDS, so we can
+						// avoid having to configure that as well
+						LoadAssignment: &api.ClusterLoadAssignment{
+							ClusterName: envoyServiceName,
+							Endpoints: []*endpoint.LocalityLbEndpoints{{
+								LbEndpoints: s.EnvoyServiceFromService(svcEndpoint, port.ServicePort),
+							}},
+						},
+						// TODO: Should we set any of these timeouts? Contour believes
+						// the IdleTimeout should be set to 60s. See here: https://github.com/projectcontour/contour/blob/2858fec20d26f56cc75a19d91b61d625a86f36de/internal/envoy/listener.go#L102-L106
+						// CommonHttpProtocolOptions: &core.HttpProtocolOptions{
+						// 	IdleTimeout:           &duration.Duration{Seconds: 60},
+						// 	MaxConnectionDuration: &duration.Duration{Seconds: 60},
+						// },
+						// TODO: Do we need / want to enable this? If yes,
```

Comment:
I don't know.

Reply:
OK, I'll leave them there as comments in case we'll need to revisit them.